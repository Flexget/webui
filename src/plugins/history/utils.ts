import { groupBy } from 'utils/array';
import { HistoryItem, GroupByFields } from './types';

export const getGroupedHistory = (items: HistoryItem[], grouping: GroupByFields) =>
  groupBy(
    items,
    grouping,
    grouping === GroupByFields.Time ? (i: string) => i.split('T')[0] : undefined,
  );
