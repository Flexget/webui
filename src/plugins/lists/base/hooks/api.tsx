import React, { createContext, useContext, ComponentType } from 'react';
import { RequestState, APIRequest } from 'core/api';
import { ContainerProviderProps } from 'unstated-next';
import { OverflowMenuProps } from 'core/layout/AppBar/OverflowMenu';
import { Option } from 'common/inputs/SelectField';
import { List, Entry } from '../types';

interface ListAPI {
  useGet: () => [RequestState, APIRequest<List[]>];
  useAdd: () => [RequestState, APIRequest<List>];
  useRemove: (listId?: number) => [RequestState, APIRequest];
}

interface EntryAPI {
  useGet: (listId: number | undefined, query: string) => [RequestState, APIRequest<Entry[]>];
  useAdd: (listId?: number) => [RequestState, APIRequest<Entry>];
  useRemove: (listId?: number, entryId?: number) => [RequestState, APIRequest];
  useRemoveBulk: (listId?: number) => [RequestState, APIRequest];
}

export interface API {
  list: ListAPI;
  entry: EntryAPI;
}

interface EntryProps<T extends Entry> {
  entry: T;
}

export interface Card<T extends Entry> {
  ActionsLeft?: ComponentType<EntryProps<T>>;
  Actions?: ComponentType<EntryProps<T>>;
}

interface ListState<T extends Entry> {
  api: API;
  card?: Card<T>;
  useMenuProps?: () => OverflowMenuProps[];
  sortByOptions: Option[];
}

const PluginContext = createContext<unknown>(null);

export const usePluginContainer = <T extends Entry>() => {
  const value = useContext(PluginContext);
  if (value === null) {
    throw new Error('Component must be wrapped with <Container.Provider>');
  }
  return value as ListState<T>;
};

export function createPluginContainer<T extends Entry, State = void>(
  useHook: (initialState?: State) => ListState<T>,
) {
  function Provider({ initialState, children }: ContainerProviderProps<State>) {
    const value = useHook(initialState);
    return <PluginContext.Provider value={value}>{children}</PluginContext.Provider>;
  }

  return { Provider, useContainer: usePluginContainer };
}
