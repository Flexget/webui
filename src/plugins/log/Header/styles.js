import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import theme from 'theme';

import { Typography, TextField } from '@material-ui/core';

const grey = css`
  color: ${theme.palette.grey[600]};
`;

export const Wrapper = styled.div`
  display: block;
  ${theme.breakpoints.up('sm')} {
    display: flex;
    align-items: baseline;
  }
`;

export const GreyIcon = styled(FontAwesomeIcon)`
  ${grey};
`;

export const GreyClickableIcon = styled(GreyIcon)`
  cursor: pointer;
`;

export const GreyType = styled(Typography)`
  ${grey};
`;

export const Spacer = styled.div`
  flex: 1;
`;

export const TextFieldWrapper = styled.div`
  display: flex;
  align-items: baseline;
`;

export const FilterField = styled(TextField)`
  margin-left: 1rem;
  margin-right: 1rem;
`;

export const MenuIcon = styled(FontAwesomeIcon)`
  margin-right: 3rem;
`;
