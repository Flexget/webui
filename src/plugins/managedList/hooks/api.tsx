import React, { createContext, useContext } from 'react';
import { RequestState, APIRequest, APIRequestCreator } from 'core/api';
import { ContainerProviderProps } from 'unstated-next';
import { List, Entry } from '../types';

type SelectedId = number | undefined;
interface ListAPI {
  get: [RequestState, APIRequest<List[]>];
  add: [RequestState, APIRequest<List>];
  remove: [RequestState, APIRequestCreator<[SelectedId]>];
}

interface EntryAPI {
  get: [RequestState, APIRequestCreator<[SelectedId, string], Entry[]>];
  add: [RequestState, APIRequestCreator<[SelectedId], Entry>];
  update: [RequestState, APIRequestCreator<[SelectedId, number], Entry>];
  remove: [RequestState, APIRequestCreator<[SelectedId, SelectedId]>];
  removeBulk: [RequestState, APIRequestCreator<[SelectedId]>];
  updateBulk: [RequestState, APIRequestCreator<[SelectedId], Entry[]>];
}

export interface API {
  list: ListAPI;
  entry: EntryAPI;
}

const PluginContext = createContext<API | null>(null);

export const usePluginContainer = () => {
  const value = useContext(PluginContext);
  if (value === null) {
    throw new Error('Component must be wrapped with <Container.Provider>');
  }
  return value;
};

export function createPluginContainer<State = void>(useHook: (initialState?: State) => API) {
  function Provider({ initialState, children }: ContainerProviderProps<State>) {
    const value = useHook(initialState);
    return <PluginContext.Provider value={value}>{children}</PluginContext.Provider>;
  }

  return { Provider, useContainer: usePluginContainer };
}
