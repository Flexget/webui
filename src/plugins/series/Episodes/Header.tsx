import React, { FC, ChangeEvent, useEffect } from 'react';
import { TablePagination, Theme } from '@material-ui/core';
import { css, ClassNames } from '@emotion/core';
import SelectField from 'common/inputs/formik/SelectField';
import { useContainer } from 'unstated-next';
import { useFormikContext } from 'formik';
import { useDebounce } from 'utils/hooks';
import { Direction } from 'utils/query';
import { EpisodeContainer } from '../hooks/episodes';
import { GetEpisodeOptions } from '../types';

interface Props {
  options: GetEpisodeOptions;
}

const wrapper = (theme: Theme) => css`
  display: flex;
  justify-content: space-between;
  padding-bottom: ${theme.typography.pxToRem(theme.spacing(2))};
  align-items: center;
`;

const container = css`
  display: flex;
  align-items: center;
  font-size: 14px;
`;

const item = (theme: Theme) => css`
  margin-left: ${theme.typography.pxToRem(theme.spacing(2))};
  margin-right: ${theme.typography.pxToRem(theme.spacing(2))};
`;

const input = css`
  font-size: inherit;
`;

const orderByOptions = [
  {
    value: Direction.Desc,
    label: 'Desc',
  },
  {
    value: Direction.Asc,
    label: 'Asc',
  },
];

const Header: FC<Props> = ({ options: { page, perPage } }) => {
  const [{ totalCount }] = useContainer(EpisodeContainer);

  const { values, submitForm, setFieldValue } = useFormikContext<GetEpisodeOptions>();

  const handleChangePerPage = (event: ChangeEvent<HTMLInputElement>) => {
    setFieldValue('perPage', parseInt(event.target.value, 10));
    setFieldValue('page', 0);
  };

  const handleChangePage = (_: unknown, p: number) => {
    setFieldValue('page', p);
  };

  const debouncedValues = useDebounce(values);

  useEffect(() => {
    submitForm();
  }, [...Object.values(debouncedValues), submitForm]); // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <>
      <ClassNames>
        {({ css: cssString }) => (
          <div css={wrapper}>
            <div css={container} />
            <div css={container}>
              <SelectField
                css={item}
                label="Order"
                name="order"
                id="order"
                size="small"
                InputProps={{ className: cssString(input) }}
                options={orderByOptions}
              />
              <TablePagination
                rowsPerPageOptions={[30, 60, 90]}
                count={totalCount}
                rowsPerPage={perPage}
                page={page}
                onChangePage={handleChangePage}
                onChangeRowsPerPage={handleChangePerPage}
                component="div"
              />
            </div>
          </div>
        )}
      </ClassNames>
    </>
  );
};

export default Header;
