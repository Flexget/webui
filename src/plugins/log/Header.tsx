import React, {
  FC,
  useState,
  useCallback,
  MouseEvent,
  useMemo,
  ComponentType,
  useEffect,
} from 'react';
import { useFormikContext } from 'formik';
import {
  Theme,
  Typography,
  IconButton,
  Menu,
  MenuItem,
  ListItemIcon,
  CircularProgress,
} from '@material-ui/core';
import { css } from '@emotion/core';
import { Spacer } from 'common/styles';
import TextField from 'common/inputs/formik/TextField';
import { PlayArrow, Stop, FilterList, MoreVert, ClearAll } from '@material-ui/icons';
import { ReadyState } from 'core/api';
import { useDebounce } from 'utils/hooks';
import { Options } from './types';

export const wrapper = (theme: Theme) => css`
  display: block;
  ${theme.breakpoints.up('sm')} {
    display: flex;
    align-items: baseline;
  }
`;

export const filterWrapper = css`
  display: flex;
  align-items: center;
`;

export const filterField = css`
  margin-left: 1rem;
  margin-right: 1rem;
`;

interface Props {
  readyState: ReadyState;
  connect: () => void;
  disconnect: () => void;
  clear: () => void;
}

interface StateOptions {
  onClick?: () => void;
  Icon: ComponentType;
  label: string;
  heading: string;
  disabled?: boolean;
}

const Header: FC<Props> = ({ readyState, connect, disconnect, clear }) => {
  const [anchorEl, setAnchorEl] = useState<HTMLButtonElement>();
  const helperText = 'Supports operators and, or, (), and "str"';

  const stateMap: Record<ReadyState, StateOptions> = useMemo(
    () => ({
      [ReadyState.Open]: {
        onClick: disconnect,
        Icon: Stop,
        label: 'Stop',
        heading: 'Streaming',
      },
      [ReadyState.Connecting]: {
        label: 'Start',
        Icon: CircularProgress,
        heading: 'Connecting',
        disabled: true,
      },
      [ReadyState.Closed]: {
        onClick: connect,
        Icon: PlayArrow,
        label: 'Start',
        heading: 'Disconnected',
      },
    }),
    [connect, disconnect],
  );

  const { onClick, Icon, heading, label, disabled = false } = stateMap[readyState];

  const handleMenuClose = useCallback(() => setAnchorEl(undefined), []);
  const handleMenuOpen = useCallback(
    (event: MouseEvent<HTMLButtonElement>) => setAnchorEl(event.currentTarget),
    [],
  );

  const { values, submitForm } = useFormikContext<Options>();

  const debouncedValues = useDebounce(values);

  useEffect(() => {
    submitForm();
  }, [...Object.values(debouncedValues), submitForm]);

  return (
    <div css={wrapper}>
      <div>
        <Typography variant="h6">Server Log</Typography>
        <Typography variant="subtitle1" color="textSecondary">
          {heading}
        </Typography>
      </div>
      <Spacer />
      <div css={filterWrapper}>
        <FilterList />
        <TextField
          css={filterField}
          id="query"
          name="query"
          label="Filter"
          helperText={helperText}
        />
        <IconButton onClick={handleMenuOpen}>
          <MoreVert />
        </IconButton>
      </div>
      <Menu id="log-menu" anchorEl={anchorEl} open={!!anchorEl} onClose={handleMenuClose}>
        <MenuItem>
          <TextField id="lines" name="lines" label="Max Lines" type="number" />
        </MenuItem>
        <MenuItem onClick={clear}>
          <ListItemIcon>
            <ClearAll />
          </ListItemIcon>
          Clear
        </MenuItem>
        <MenuItem onClick={onClick} disabled={disabled}>
          <ListItemIcon>
            <Icon />
          </ListItemIcon>
          {label}
        </MenuItem>
      </Menu>
    </div>
  );
};

export default Header;
