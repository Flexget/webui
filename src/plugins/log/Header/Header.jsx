import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Menu, MenuItem, TextField, Typography, IconButton, ListItemIcon } from '@material-ui/core';
import { FilterList, MoreVert, Stop, PlayArrow, ClearAll } from '@material-ui/icons';
import { Wrapper, Spacer, TextFieldWrapper, FilterField } from './styles';

const ENTER_KEY = 13;

class Header extends Component {
  static propTypes = {
    start: PropTypes.func.isRequired,
    stop: PropTypes.func.isRequired,
    connected: PropTypes.bool.isRequired,
    clearLogs: PropTypes.func.isRequired,
  };

  state = {
    open: false,
    lines: '200',
    query: '',
  };

  componentDidMount() {
    this.reload();
  }

  componentWillUnmount() {
    const { stop } = this.props;
    stop();
  }

  reload = () => {
    const { connected, start, stop } = this.props;
    const { lines, query } = this.state;

    if (connected) {
      stop();
    }
    start({ lines, query });
  };

  clearLogs = () => this.props.clearLogs();

  handleLines = event => this.setState({ lines: event.target.value });

  handleQuery = event => this.setState({ query: event.target.value });

  handleKeyPress = event => event.which === ENTER_KEY && this.reload();

  handleRequestClose = () =>
    this.setState({
      open: false,
      anchorEl: undefined,
    });

  handleMenuClick = event =>
    this.setState({
      open: true,
      anchorEl: event.currentTarget,
    });

  render() {
    const { connected, stop } = this.props;
    const { anchorEl, open, query, lines } = this.state;
    const helperText = 'Supports operators and, or, (), and "str"';
    const { onClick, Icon, text } = connected
      ? {
          onClick: stop,
          Icon: Stop,
          text: 'Stop',
        }
      : {
          onClick: this.reload,
          Icon: PlayArrow,
          text: 'Start',
        };

    return (
      <Wrapper>
        <div>
          <Typography variant="h6">Server Log</Typography>
          <Typography variant="subtitle1" color="textSecondary">
            {connected ? 'Streaming' : 'Disconnected'}
          </Typography>
        </div>
        <Spacer />
        <TextFieldWrapper>
          <FilterList />
          <FilterField
            id="filter"
            label="Filter"
            value={query}
            onChange={this.handleQuery}
            inputProps={{
              onKeyPress: this.handleKeyPress,
            }}
            helperText={helperText}
          />
          <IconButton onClick={this.handleMenuClick}>
            <MoreVert />
          </IconButton>
        </TextFieldWrapper>
        <Menu id="log-menu" anchorEl={anchorEl} open={open} onClose={this.handleRequestClose}>
          <MenuItem>
            <TextField
              id="lines"
              label="Max Lines"
              value={lines}
              onChange={this.handleLines}
              type="number"
              inputProps={{
                onKeyPress: this.handleKeyPress,
              }}
            />
          </MenuItem>
          <MenuItem onClick={this.clearLogs}>
            <ListItemIcon>
              <ClearAll />
            </ListItemIcon>
            Clear
          </MenuItem>
          <MenuItem onClick={onClick}>
            <ListItemIcon>
              <Icon />
            </ListItemIcon>
            {text}
          </MenuItem>
        </Menu>
      </Wrapper>
    );
  }
}

export default Header;
