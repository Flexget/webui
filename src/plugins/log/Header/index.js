import { connect } from 'react-redux';
import actions from 'plugins/log/state/actions';
import Header from './Header';

export function mapStateToProps({ log }) {
  return {
    connected: log.connected,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    start: payload => dispatch(actions.connect.request(payload)),
    stop: () => dispatch(actions.disconnect()),
    clearLogs: () => dispatch(actions.clear()),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Header);
export { Header };
