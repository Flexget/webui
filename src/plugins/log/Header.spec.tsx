import React from 'react';
import { cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Formik } from 'formik';
import { themed } from 'utils/tests';
import { ReadyState } from 'core/api';
import Header from './Header';

describe('plugins/log/Header', () => {
  afterEach(() => {
    cleanup();
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        themed(
          <Formik initialValues={{ query: '', lines: 400 }} onSubmit={jest.fn()}>
            <Header
              connect={jest.fn()}
              readyState={ReadyState.Open}
              disconnect={jest.fn()}
              clear={jest.fn()}
            />
          </Formik>,
        ),
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
