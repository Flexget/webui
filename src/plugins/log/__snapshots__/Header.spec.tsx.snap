// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/log/Header renders correctly 1`] = `
.emotion-0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div
  css={[Function]}
>
  <div>
    <h6
      className="MuiTypography-root MuiTypography-h6"
    >
      Server Log
    </h6>
    <h6
      className="MuiTypography-root MuiTypography-subtitle1 MuiTypography-colorTextSecondary"
    >
      Streaming
    </h6>
  </div>
  <div
    className="emotion-0"
  />
  <div
    css={
      Object {
        "map": undefined,
        "name": "1sg2lsz",
        "next": undefined,
        "styles": "
  display: flex;
  align-items: center;
",
        "toString": [Function],
      }
    }
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
      />
    </svg>
    <div
      className="MuiFormControl-root MuiTextField-root"
      css={
        Object {
          "map": undefined,
          "name": "1o7zzk8",
          "next": undefined,
          "styles": "
  margin-left: 1rem;
  margin-right: 1rem;
",
          "toString": [Function],
        }
      }
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
        htmlFor="query"
        id="query-label"
      >
        Filter
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-describedby="query-helper-text"
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          id="query"
          name="query"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
      <p
        className="MuiFormHelperText-root"
        id="query-helper-text"
      >
        Supports operators and, or, (), and "str"
      </p>
    </div>
    <button
      className="MuiButtonBase-root MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
